align_system/algorithms/outlines_adm.py

```diff
@@ -349,24 +361,73 @@ def top_level_choose_action(self,
+            prompt = action_selection_prompt(scenario_description, shuf)
             dialog = [{'role': 'system', 'content': negative_system_prompt}]
             dialog.extend(negative_icl_examples)
             dialog.append({'role': 'user', 'content': prompt})
-
             negative_dialogs.append(dialog)
 
+        return {"scenario_description": scenario_description,
+                "choices": choices,
+                "positive_system_prompt": positive_system_prompt,
+                "negative_system_prompt": negative_system_prompt,
+                "positive_dialogs": positive_dialogs,
+                "negative_dialogs": negative_dialogs}
+
+    def top_level_choose_action(self,
+                                scenario_state,
+                                available_actions,
+                                alignment_target,
+                                num_positive_samples=1,
+                                num_negative_samples=0,
+                                generator_batch_size=5,
+                                kdma_descriptions_map='align_system/prompt_engineering/kdma_descriptions.yml',
+                                reasoning_max_length=512,
+                                generator_seed=-1,
+                                max_generator_tokens=-1,
+                                shuffle_choices=True,
+                                **kwargs):
+        if self.baseline and num_negative_samples > 0:
+            raise RuntimeError("No notion of negative samples for baseline run")
+        if self.baseline and "incontext" in kwargs and kwargs["incontext"]["number"] > 0:
+            raise RuntimeError("No notion of incontext examples for baseline run")
+
+        dialogs_data = OutlinesTransformersADM.get_dialogs(
+            scenario_state,
+            available_actions,
+            alignment_target,
+            num_positive_samples,
+            num_negative_samples,
+            kdma_descriptions_map,
+            shuffle_choices,
+            baseline=self.baseline,
+            scenario_description_template=self.scenario_description_template,
+            action_selection_prompt_template=self.action_selection_prompt_template,
+            baseline_system_prompt=self.baseline_system_prompt,
+        )
+        choices = dialogs_data["choices"]
+        positive_dialogs = dialogs_data["positive_dialogs"]
+        negative_dialogs = dialogs_data["negative_dialogs"]
+
         # Need to set the whitespace_pattern to prevent the state
         # machine from looping indefinitely in some cases, see:
         # https://github.com/outlines-dev/outlines/issues/690#issuecomment-2102291934
         generator = outlines.generate.json(
             self.model,
-            action_choice_json_schema(json.dumps(choices)),
+            action_choice_json_schema(json.dumps(choices), reasoning_max_length),
             sampler=self.sampler,
             whitespace_pattern=r"[ ]?")
 
+        if max_generator_tokens >= 0:
```

Comment:
This is an interesting approach, normally I would set the default in the kwargs to `None` and then if it's not `None` pass the value along as-is; but I guess in this case (setting the default to `-1`) you get to do some semantic validation on the value at the same time.  Is that kind of the intention or is this just a tomato vs tomato.

Reply:
The -1 is a gross hack, and I regret it already =)

Reply:
I wouldn't say it's a hack, just curious about the tradeoffs; I'm content to leave it as is.